libraries/lib-music-information-retrieval/MirDsp.cpp

```diff
@@ -0,0 +1,138 @@
+#include "MirDsp.h"
+#include "FFT.h"
+#include "MirAudioReader.h"
+#include "MirTypes.h"
+#include "MirUtils.h"
+#include "StftFrameProvider.h"
+
+#include <cassert>
+#include <cmath>
+#include <numeric>
+
+namespace MIR
+{
+namespace
+{
+float GetNoveltyMeasure(
+   const std::vector<float>& prevPowSpec, const std::vector<float>& powSpec)
+{
+   auto k = 0;
+   return std::accumulate(
+      powSpec.begin(), powSpec.end(), 0., [&](float a, float mag) {
+         // Half-wave-rectified stuff
+         return a + std::max(0.f, mag - prevPowSpec[k++]);
+      });
+}
+
+std::vector<float>
+GetMovingAverage(const std::vector<float>& x, double sampleRate)
+{
+   constexpr auto smoothingWindowDuration = 0.2;
+   // An odd number.
+   const int M = std::round(smoothingWindowDuration * sampleRate / 4) * 2 + 1;
+   const auto window = GetNormalizedHann(2 * M + 1);
+   auto n = 0;
+   std::vector<float> movingAverage(x.size());
+   std::transform(x.begin(), x.end(), movingAverage.begin(), [&](float) {
+      auto y = 0.;
+      for (auto i = -M; i <= M; ++i)
+      {
+         auto k = n + i;
+         if (k < 0)
+            k += x.size();
+         else if (k >= x.size())
+            k -= x.size();
+         y += x[k] * window[i + M];
+      }
+      ++n;
+      constexpr auto smoothingThreshold = 1.5;
```

Comment:
I don't understand this multiplier.

Reply:
It's a poorly documented yet important multiplier.
The moving average of the raw ODF will be subtracted from it to yield the final ODF, negative results being set to 0. This is to remove noise of small ODF peaks before the method's quantization step.
Changing this value can impact dramatically the performance of the classifier.
I'm still not completely happy with the ODF. In some cases it can be very noisy in comparison with our perception of where onsets are. I'm exploring better solutions, and if I do find one, this smoothing may become obsolete.
Still, adding a comment.